dagshub/data_engine/model/dataset_state.py

```diff
@@ -0,0 +1,62 @@
+import json
+from dataclasses import dataclass
+from typing import Dict, Optional, List, Union
+
+from dagshub.data_engine.client.models import DatasetResult
+from dagshub.data_engine.model.query import DatasourceQuery
+
+
+@dataclass
+class DatasetState:
```

Comment:
This one is also cached inside of `Datasource` (`ds.assigned_dataset`), so datasource can know the query of the dataset and other info relating to that.
I think it'd be weird to have a `DatasourceSerializedState` be a field in `Datasource`

Reply:
Also the Query in this one is not really serializable.
I guess I can make the DatasourceQuery object serializable/deserializable by itself. That'd probably simplify things a lot 🤔

Reply:
will cut a couple serialization sub-structs

Reply:
Took a glance, I'll have to make `DatasourceQuery` a dataclass, which it isn't right now, which might be going into abuse territory, because there's also the whole hidden state with the tree, which i DON'T want to be serializing/deserializing.

Maybe I should write a custom [JSONEncoder](https://docs.python.org/3/library/json.html#json.JSONEncoder) and use that throughout Data Engine files, but that interface is pretty clunky imo

Reply:
I'm a bit confused on how the query isn't serializeable - you obviously do serialize it.
Not that any of it is critical for merging the PR...

Reply:
It is de/serializable, but only in the context of "I need to call a whole another function to actually do that". I don't know of a neat way to change the signature in DatasourceSerializedState to DatasourceQuery and say "hey actually you need to use this function to serialize, and this to deserialize"

Reply:
Made `DatasourceSerializedState` actually have the `DatasourceQuery` in it in the end :)